### DISCUSSION TITLE
ESP3D V3 - Sensors bug

### DISCUSSION THREAD

**Discussion by Stephano2000** [luc-github/ESP3D#858] [Questions]
Is the sensors feature working in ESP3D V3?
I am using an ESP32-CAM with a DHT11 sensor connected to GPIO16. Once I change  the sensor type from None to DHT11 in Web UI/Settings/Features, the camera panel becomes blank, and the sensor data isn't shown anywhere. I tested the hardware setup with a DHT11 test sketch, and the sensor works fine.

Moreover, ESP210.cpp has a syntax error that prevents compilation (patch attached)

[ESP210.diff.txt](https://github.com/luc-github/ESP3D/files/10561706/ESP210.diff.txt)

**Comment by luc-github** [comment]
you cannot use GPIO 16/17 they are used by PSRAM which is used by camera

indeed there is a typo error - now fixed than you https://github.com/luc-github/ESP3D/commit/5ce1d16479c6f2b5d725badbe5e8ae99ccddce80

**Comment by Stephano2000** [comment]
You're right regarding GPIO16. I missed the fact that it had 2 labels on the schematic.
As there is no free GPIOs, I used GPIO12, which is shared with the SD card, but since i have not enabled the SD card in ESP3D, i suppose that it should be ok.
In fact the camera works fine now, but the sensor output is still nowhere to be seen on the dashboard.

**Reply by luc-github** [reply]
I suggest you to not use GPIO 12 - it is strapping pin , I used 13 with my DHT 22 (same driver as DHT11)

and it works on my side

![image](https://user-images.githubusercontent.com/8822552/216330966-93fe4e73-24a0-4cbd-8cbe-e67cd7b568cc.png)

![image](https://user-images.githubusercontent.com/8822552/216331432-2b4d9d75-3c5b-4367-bcae-1b0775209951.png)


![image](https://user-images.githubusercontent.com/8822552/216331178-df8910f1-8772-445a-b489-050f6af07574.png)

**Reply by luc-github** [reply]
But it seems I have a glitch in the configurator the DHT does not need `SENSOR_ADDR` actually - will check that tomorrow

**Comment by Stephano2000** [comment]
> I suggest you to not use GPIO 12 - it is strapping pin , I used 13 with my DHT 22 (same driver as DHT11)
> and it works on my side

I don't see on the ESP)32-CAM schematic anything regarding GPIO12 being other than a SD card data pin.

Anyway I tried GPIO13 as you suggested (though your configuration.h shows pin 34), but still nothing concerning sensors on the dashboard, although the settings page shows DHT11 with a 1000ms polling interval.

Also from the terminal:
[ESP210]
ms, value=DISCONNECTED
[ESP210]type=DHT11
ok


![Screenshot from 2023-02-02 15-37-12](https://user-images.githubusercontent.com/3524476/216340244-e0d3edfe-e6ba-426f-9610-b442be920fe0.png)
![Screenshot from 2023-02-02 15-37-02](https://user-images.githubusercontent.com/3524476/216340327-a91d0c14-7c27-4224-98d5-baf0e6c33c00.png)
![Screenshot from 2023-02-02 15-36-44](https://user-images.githubusercontent.com/3524476/216340352-64cedeea-9886-4d8e-acf1-aff2c944c1f3.png)

**Reply by luc-github** [reply]
1- here the reference FYI https://randomnerdtutorials.com/esp32-pinout-reference-gpios/
2 - check the right place in dev tool - the websockets messages - not the get

**Reply by Stephano2000** [reply]
1- Not sure i understand, or maybe there is a forgotten attachment.
2- Websocket:
![Screenshot from 2023-02-02 15-52-27](https://user-images.githubusercontent.com/3524476/216343472-1e47ed99-cd81-4297-b15a-3bb8d1d0ecee.png)

**Reply by luc-github** [reply]
additionnally DTH take time to collect data - 1000ms is too short https://github.com/luc-github/ESP3D/blob/2.1.x/libraries/DHT_sensor_library_for_ESPx/examples/DHT_ESP32/DHT_ESP32.ino#L103

2000ms is the minimum IMHO

**Reply by luc-github** [reply]
what browser are you using ?

**Reply by Stephano2000** [reply]
Firefox

**Reply by luc-github** [reply]
![image](https://user-images.githubusercontent.com/8822552/216345718-2fe8d60a-dd04-411b-8fc9-b3dc11809285.png)

**Reply by luc-github** [reply]
As your command show disconnected check your wires,  I had some issue with my dupont wires which are loose - it may be your issue too

Checking your [ESP210] I see output is incorrect 
I have just fixed it 
should be :
`entry 0x400805e4
M117 Setup wifi
M117 Connecting
M117 192.168.2.177
M117 192.168.2.177
[ESP210]
;type=DHT11, interval=5000ms, value=25.2[C] 67.3[%]
`

**Comment by Stephano2000** [comment]
Are you using an ESP32-CAM or ESP32 dev board or module?
I changed the DHT11 interval to 5000ms.

![Screenshot from 2023-02-02 16-12-59](https://user-images.githubusercontent.com/3524476/216348569-c9df7672-b41c-41af-a41a-d02e933d6552.png)

**Reply by Stephano2000** [reply]
The sensor works when i upload a DHT11 test sketch to the ESP32-CAM

**Reply by luc-github** [reply]
No I do not have the esp32cam in hand now - I test with standard esp32
can you share your configure.h ?

**Reply by Stephano2000** [reply]
configuration.h
[configuration.h.txt](https://github.com/luc-github/ESP3D/files/10569881/configuration.h.txt)

**Reply by luc-github** [reply]
I see nothing wrong - does issue still happen if you disable camera or camera refresh ?

**Reply by Stephano2000** [reply]
I deleted the camera panel, and still no sensor stuff. Should i delete the camera from configuration.h altogether?

Btw, are you using ESP3D V3 git? It had a typo and doesn't compile, so maybe there is some other bug?

**Reply by luc-github** [reply]
>are you using ESP3D V3 git

Yes I am using latest git , I just pushed fix for [ESP210]

> Should i delete the camera from configuration.h

yes please

**Reply by Stephano2000** [reply]
Still the same without CAM in configuration.h. [ESP210] now outputs some more data, but still shows disconnected.

**Reply by Stephano2000** [reply]
Debug output:
[ESP3D][sensor.cpp:57] begin(): Sensor Begin
[ESP3D][sensor.cpp:74] begin(): Sensor Device created
[ESP3D][sensor.cpp:76] begin(): Sensor 1
[ESP3D][dht.cpp:63] begin(): DHT PIN 13
[ESP3D][dht.cpp:69] begin(): DHT ok
...
[ESP3D][dht.cpp:131] GetData(): T nan H nan
[ESP3D][dht.cpp:142] GetData(): No valid data

**Reply by luc-github** [reply]
the device creation looks ok and because I use library api and it is so simple I doubt it come from my code more over I cannot duplicate your issue.

* how do you compile/flash ? arduino/Platformio ?
 if arduino ide what esp32 core version are you using ?
 
* try using without PSRAM, I do not think I have tested with PSRAM
use esp32dev in platformio or disable PSRAM in arduino IDE

* what version of DHT lib are you using ? the one I provide or another one ?

**Reply by Stephano2000** [reply]
- I tried both arduino IDE (ESP32 core 2.0.6) and platformio
- I tested using esp32dev in platformio, and the issue is still there
- I am using the DHT library you provided with ESP3D V3 (for arduino ide)
- I tested the DHT11 on the ESP32-CAM with a simple sketch (attached) using both arduino ide and platformio and it works

[DHT_ESP32CAM-Test.ino.txt](https://github.com/luc-github/ESP3D/files/10576182/DHT_ESP32CAM-Test.ino.txt)

**Reply by Stephano2000** [reply]
Maybe you can build a firmware file that i can test on the ESP32-CAM?

**Reply by luc-github** [reply]
I do not use the API I see in your test code, I use the API present in default sample : https://github.com/luc-github/ESP3D/blob/3.0/libraries/DHT_sensor_library_for_ESPx-1.0.6/examples/DHT_Test/DHT_Test.ino

in https://github.com/luc-github/ESP3D/blob/3.0/esp3d/src/modules/sensor/dht.cpp

**Reply by Stephano2000** [reply]
It's the same API, but i used getTempAndHumidity instead of getTemperature and getHumidity.
Anyway, DHT_Test.ino works flawlessly on the ESP32-CAM (GPIO13):
OK	 H:77.0%	T :16.0C		
OK	 H:77.0%	T :16.0C		
OK	 H:79.0%	T :16.0C

**Reply by luc-github** [reply]
Then I have no idea why it works on my side and not on yours

**Reply by Stephano2000** [reply]
I finally got it working albeit through a hack, without getting to the root cause yet.
In dht.cpp, I added a supposedly redundant setup call in DHTSensorDevice::GetData():
        dht_device->setup(15, DHTesp::DHT11); // hack
        float temperature = dht_device->getTemperature();
        float humidity= dht_device->getHumidity();
        ...
The setup function sets the pin and model, resets the timer, and reads the sensor. (in autodetect mode only) How to debug further?
![Screenshot from 2023-02-03 16-10-44](https://user-images.githubusercontent.com/3524476/216624506-a9701634-a66b-44e8-a3cc-bdf29c761352.png)

**Reply by luc-github** [reply]
you also changed pin ?

**Reply by Stephano2000** [reply]
i am using GPIO15 even in configuration.h

**Reply by Stephano2000** [reply]
I added debug code in dht library setup (1st line) to show what pin and model are currently used.
After the 1st dht setup, something is changing the pin to 112 and the model to some wild value.
Then the values persist.

**Comment by Stephano2000** [comment]
This is the debug output. I added debug info at the entry and exit of dht setup function.
Notice how the pin and model are setup correctly at the beginning, then the 2nd call to setup (that i added) shows different values.
[output.txt](https://github.com/luc-github/ESP3D/files/10579828/output.txt)

**Reply by Stephano2000** [reply]
Also why would the sensor model be set to 2 in the 1st call to dht setup while  the sensor is a DHT11?
[ESP3D][sensor.cpp:57] begin(): Sensor Begin
[ESP3D][sensor.cpp:74] begin(): Sensor Device created
[ESP3D][sensor.cpp:76] begin(): Sensor 1
[ESP3D][dht.cpp:63] begin(): DHT PIN 15
 [start pin]:5 start model:1073451288
 [end pin]:15 end model:2

**Reply by Stephano2000** [reply]
Changing the model in dht_device->setup call from SENSOR_TYPE[dhttype] to DHTesp::DHT11 is enough to let the sensor work.
Maybe the casting should be changed to SENSOR_TYPE[dhttype-1] as SENSOR_TYPE[1] is DHT22?
```
--- dht.cpp
+++ dht.cpp
@@ -61,7 +61,7 @@
         return false;
     }
     log_esp3d("DHT PIN %d",ESP3D_SENSOR_PIN);
-    dht_device->setup(ESP3D_SENSOR_PIN, SENSOR_TYPE[dhttype]);
+    dht_device->setup(ESP3D_SENSOR_PIN, DHTesp::DHT11);
     if (strcmp(dht_device->getStatusString(), "OK")!=0) {
         log_esp3d("No valid dht status: %d,  %s",dht_device->getStatus(), dht_device->getStatusString());
         return false;
```

**Reply by luc-github** [reply]
Hi thanks to your debug I have found that the setup of sensor used wrong value, it used index in array instead of ID itself, the result is it used always DHT22 instead of DHT11, it also bring me to several report issue about sensor name ... orz 
So on my side it was always working on my DHT22 even I put on DHT11 that was actually ... a DHT22

Please give a try to latest git - thank to your debug issue should be fixed

**Reply by Stephano2000** [reply]
Tested latest git and it works.
It was a stroke of luck that it worked for the DHT22. It was rolling over to 0, which is the AUTO_DETECT of the DHT library.

Please move this question to the closed bugs/issues section.

**Reply by luc-github** [reply]
yes and it took me some time to understand where the issue come from even with your debug
